fusesoc/build/vivado.py

```diff
@@ -0,0 +1,170 @@
+
+""" Vivado Backend
+
+The Vivado backend executes Xilinx Vivado to build systems and program the FPGA.
+
+The backend defines the following section:
+
+    [vivado]
+    part = <part name> # Format <family><device><package>-<speedgrade>
+    hw_device = <device name> # Format <family><device>_0
+
+A core (usually the system core) can add the following files:
+
+- Standard design sources
+
+- Constraints: Supply xdc files with file_type=xdc
+
+- IP: Supply the IP core xdi file with file_type=xdi and other files (like .prj)
+      as file_type=data
+"""
+class Vivado(Backend):
+    """ Define the toolname. It is used by FuseSoC to select the backend"""
+    TOOL_NAME = 'vivado'
+
+    """ The constructor of the backend"""
+    def __init__(self, system):
+        # Do the general backend initialization
+        super(Vivado, self).__init__(system)
+        # Set work root. That is the name of the output path for build files
+        self.work_root = os.path.join(self.build_root, 'bld-{}'.format(self.TOOL_NAME))
+
+    """ Configuration is the first phase of the build
+
+    In the vivado backend the project TCL is written and all files are copied
+    """
+    def configure(self, args):
+        super(Vivado, self).configure(args)
+        self._write_project_tcl_file()
+
+    """ Write the project tcl file
+
+    This writes the project tcl file. It first collects all sources, ip and contraints
+    and then writes them to the tcl file along with the build steps.
+    """
+    def _write_project_tcl_file(self):
+        ip = []      # IP descriptions (xci files)
+        constr = []  # Constraints (xdc files)
+        verilog = [] # (System) Verilog files
+        vhdl = []    # VHDL files
+
+        # Get the synthesis files and files specific to vivado
+        (src, incdirs) = self._get_fileset_files(['vivado', 'synth'])
+        for s in src:
+            if s.file_type == 'xci':
+                ip.append(s.name)
+            elif s.file_type == 'xdc':
+                constr.append(s.name)
+            elif s.file_type.startswith('verilogSource'):
+                verilog.append(s.name)
+            elif s.file_type.startswith('systemVerilogSource'):
+                verilog.append(s.name)
+            elif s.file_type.startswith('vhdlSource'):
+                vhdl.append(s.name)
+
+        # Write the formatted string to the tcl file
+        tcl_file = open(os.path.join(self.work_root, self.system.name+".tcl"), 'w')
+        tcl_file.write(PROJECT_TCL_TEMPLATE.format(
+            design       = self.system.name,
+            part         = self.system.backend.part,
+            bitstream    = os.path.join(self.work_root, self.system.name+'.bit'),
+            incdirs      = ' '.join(incdirs),
+            ip_files     = '\n'.join(['read_ip '+s for s in ip]),
+            src_files    = '\n'.join(['read_verilog '+s for s in verilog]+
+                                     ['read_vhdl '+s for s in vhdl]),
```

Comment:
read_vhdl must be able to set the library too. Might be required for sv files too. See https://github.com/olofk/fusesoc/commit/0a3e895fb079fbe04d85a12cde9a2f12c5fd1c83 for reference how this is done in the ISE backend